Delete a collection - UI portion

bobsilverberg opened the pull request:
Fixes #4195 

Depends on #5161 

This adds a "Delete this collection" button to the Collection screen, both in display and edit mode, which only appears if the user has access to delete the collection. For now I have styled it similar to the "Delete this picture" button on the User Profile.

Detail screen:

<img width="481" alt="screenshot 2018-06-04 13 30 18" src="https://user-images.githubusercontent.com/142755/40932170-7252477e-67fb-11e8-91f3-f3fe38a35716.png">

With confirmation:

<img width="467" alt="screenshot 2018-06-04 13 31 57" src="https://user-images.githubusercontent.com/142755/40932238-b3f32158-67fb-11e8-9c8b-5fa4c6a7dd13.png">

Edit screen:

<img width="472" alt="screenshot 2018-06-04 13 30 48" src="https://user-images.githubusercontent.com/142755/40932189-831925e6-67fb-11e8-82c0-270e54533125.png">

With confirmation:

<img width="469" alt="screenshot 2018-06-04 13 32 32" src="https://user-images.githubusercontent.com/142755/40932267-c8a5a332-67fb-11e8-831d-9b9f9edafeea.png">

bobsilverberg commented:
I see that #5161 was approved, so I'll merge that and rebase this now, to make reviewing the PR easier.

willdurand commented on tests/unit/amo/components/TestCollection.js:
```diff
@@ -886,6 +947,43 @@ describe(__filename, () => {
+    dispatchSpy.reset();
+
+    // This emulates a user clicking the delete button and confirming.
+    wrapper.instance().onDelete(
```
In my opinion, we should not use `instance()` that much, and especially not here. I'd rather find the `ConfirmButton` and use `.prop` to get the handler:

```js
const onDelete = wrapper.find(ConfirmButton).prop('onConfirm');
onDelete(createFakeEvent({ ... }));
```

I consider the `onDelete` method "internal" and therefore "private". If we rename this method for some reason, we'll have to update this test case, even if the behavior has not changed. If we stick to the props (the public interface/API in React), then we won't have to update this test case.

bobsilverberg replied:
Very good point, thanks for the suggestion.

willdurand commented on src/amo/sagas/collections.js:
```diff
@@ -338,6 +339,10 @@ export function* deleteCollection({
       errorHandlerId: errorHandler.id,
       username,
     }));
```
So it seems the `fetchUserCollectionsAction` does not remove the old/existing collections for the current logged-in user. It means deleting a collection and clicking the "back" button will load the previously deleted collection. I am not sure that's what we want.

bobsilverberg replied:
I think the issue is actually with `loadUserCollections`, not `fetchUserCollections` in that it loads all the newly found collections, but adds them to any existing collections. I believe that is because of this line: https://github.com/mozilla/addons-frontend/blob/master/src/amo/reducers/collections.js#L968. If we were to remove that, then `loadUserCollections` would always load a fresh set of collections, which seems like what we want to do. Do you have any idea why the code might have been written that way to begin with? When might we want to just add to an existing set of user collections?

willdurand replied:
The line 968 of the `collections` reducer modifies a `userCollections` attribute in the state, but when I load a collection page directly, this attribute is an empty object:

![screen shot 2018-06-05 at 19 17 09](https://user-images.githubusercontent.com/217628/40991764-16fc9532-68f5-11e8-97db-c322517ffe64.png)

It is only after I delete the collection that this `userCollections` attribute gets populated.

willdurand replied:
Isn't it `loadCollectionIntoState()` that always keep adding data to the state?

bobsilverberg replied:
Sorry, @willdurand, I am getting confused by this discussion. I'm now unclear about what you think the issue actually is. Is it as simple as making sure that a collection gets unloaded from state when we delete it, or is there more to it than that?

In either case, do you think we can make this a follow-up issue because it doesn't really interfere with the delete feature and only represents an annoyance for an edge-case?

willdurand replied:
The issue is that we don't unload a deleted collection. I think it could be done in this patch (because the diff is rather small and also because I noticed this bug while reviewing this patch). Let me know if you prefer a follow-up issue though.

bobsilverberg replied:
Ok, I've pushed a commit which unloads the collection after deleting it. Could you please take another look?

bobsilverberg commented on src/amo/components/Collection/styles.scss:
```diff
@@ -86,3 +86,11 @@
 }
+
+.Collection-delete-button {
+  margin: 12px 0;
```
I changed it to `margin-top: 16px;` because I think we tend to go with factors of 4. Let me know if that seems okay.

willdurand replied:
that's perfect!

bobsilverberg commented:
Thanks for the review, @willdurand. I have addressed all of the comments except for the one about the collection not being removed from state, to which I've added a response and some follow-up questions. I think in any case that could be treated as a follow-up issue, if you think that's okay. This is ready for another review.

willdurand commented on src/amo/components/Collection/index.js:
```diff
@@ -263,6 +287,23 @@ export class CollectionBase extends React.Component<Props> {
+    if (!hasEditPermission) {
+      return null;
+    }
+    return (
```
nit: maybe a blank line before the `return` statement would be nice for readability.

willdurand commented on src/amo/sagas/collections.js:
```diff
@@ -334,10 +335,17 @@ export function* deleteCollection({
     yield call(api.deleteCollection, params);
 
+    // Unload the collection from state.
+    yield put(deleteCollectionBySlug(slug));
```
it is definitely a follow-up issue, but having an action named `deleteCollection` that calls `deleteCollectionBySlug` is super weird. I was like "oh shhhh, I missed this `deleteCollection` action when reviewing #5161" but actually, `deleteCollection` and `deleteCollectionBySlug` have different purposes.

Can we rename `deleteCollectionBySlug` to `unloadCollectionBySlug`?

bobsilverberg replied:
Good idea. I will file a follow-up.

bobsilverberg replied:
I opened https://github.com/mozilla/addons-frontend/issues/5238
